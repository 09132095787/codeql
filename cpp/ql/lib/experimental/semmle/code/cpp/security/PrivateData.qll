/**
 * DEPRECATED: use semmle.code.cpp.security.PrivateData instead.
 */

import semmle.code.cpp.security.PrivateData
