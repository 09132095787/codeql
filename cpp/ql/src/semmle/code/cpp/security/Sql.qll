/**
 * This file provides classes for working with various SQL libraries and frameworks.
 */

private import cpp
private import semmle.code.cpp.models.interfaces.FunctionInputsAndOutputs

/**
 * An abstract class that represents SQL parameters and escaping functions.
 *
 * To add support for a new SQL framework, extend this class with
 * a subclass whose characteristic predicate is a unique singleton string.
 * For example, write
 *
 *  ```ql
 * class MySqlFunctionality extends SqlFunctionality {
 *   MySqlFunctionality() { this = "MySqlFunctionality" }
 *   // Override `getAnSqlParameter`.
 *   // Optionally override `getAnEscapedParameter`.
 * }
 * ```
 */
abstract class SqlFunctionality extends string {
  bindingset[this]
  SqlFunctionality() { any() }

  /**
   * Holds if `input` to the function `func` represents data that is passed to an SQL server.
   */
  abstract predicate getAnSqlParameter(Function func, FunctionInput input);

  /**
   * Holds if the `output` from `func` escapes the SQL input `input` such that is it safe to pass to
   * an SQL server.
   */
  predicate getAnEscapedParameter(Function func, FunctionInput input, FunctionOutput output) {
    none()
  }
}

private class MySqlFunctionality extends SqlFunctionality {
  MySqlFunctionality() { this = "MySqlFunctionality" }

  override predicate getAnSqlParameter(Function func, FunctionInput input) {
    func.hasName(["mysql_query", "mysql_real_query"]) and
    input.isParameterDeref(1)
  }
}

private class SqLite3Functionality extends SqlFunctionality {
  SqLite3Functionality() { this = "SqLite3Functionality" }

  override predicate getAnSqlParameter(Function func, FunctionInput input) {
    func.hasName("sqlite3_exec") and
    input.isParameterDeref(1)
  }
}

private module PostgreSql {
  private predicate pqxxTransactionSqlArgument(string function, int arg) {
    function = "exec" and arg = 0
    or
    function = "exec0" and arg = 0
    or
    function = "exec1" and arg = 0
    or
    function = "exec_n" and arg = 1
    or
    function = "exec_params" and arg = 0
    or
    function = "exec_params0" and arg = 0
    or
    function = "exec_params1" and arg = 0
    or
    function = "exec_params_n" and arg = 1
    or
    function = "query_value" and arg = 0
    or
    function = "stream" and arg = 0
  }

  private predicate pqxxConnectionSqlArgument(string function, int arg) {
    function = "prepare" and arg = 1
  }

  private predicate pqxxTransationClassNames(string className, string namespace) {
    namespace = "pqxx" and
    className in [
        "dbtransaction", "nontransaction", "basic_robusttransaction", "robusttransaction",
        "subtransaction", "transaction", "basic_transaction", "transaction_base", "work"
      ]
  }

  private predicate pqxxConnectionClassNames(string className, string namespace) {
    namespace = "pqxx" and
    className in ["connection_base", "basic_connection", "connection"]
  }

  private predicate pqxxEscapeArgument(string function, int arg) {
    arg = 0 and
    function in ["esc", "esc_raw", "quote", "quote_raw", "quote_name", "quote_table", "esc_like"]
  }

  class PostgreSqlFunctionality extends SqlFunctionality {
    PostgreSqlFunctionality() { this = "PostgreSqlFunctionality" }

    override predicate getAnSqlParameter(Function func, FunctionInput input) {
      exists(int argIndex, UserType t |
        func.getDeclaringType() = t and
        // transaction exec and connection prepare variations
        (
          pqxxTransationClassNames(t.getName(), t.getNamespace().getName()) and
          pqxxTransactionSqlArgument(func.getName(), argIndex)
          or
          pqxxConnectionClassNames(t.getName(), t.getNamespace().getName()) and
          pqxxConnectionSqlArgument(func.getName(), argIndex)
        ) and
        input.isParameterDeref(argIndex)
      )
    }

    override predicate getAnEscapedParameter(
      Function func, FunctionInput input, FunctionOutput output
    ) {
      exists(int argIndex, UserType t |
        func.getDeclaringType() = t and
        // transaction and connection escape functions
        (
          pqxxTransationClassNames(t.getName(), t.getNamespace().getName()) or
          pqxxConnectionClassNames(t.getName(), t.getNamespace().getName())
        ) and
        pqxxEscapeArgument(func.getName(), argIndex) and
        input.isParameterDeref(argIndex) and
        output.isReturnValueDeref()
      )
    }
  }
}

private import PostgreSql
