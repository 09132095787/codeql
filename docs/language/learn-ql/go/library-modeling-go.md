# Modeling Go libraries for CodeQL

CodeQL does not examine the source code for external packages. To track the flow of untrusted data through them you need to create a model of the library. Existing models can be found in `ql/src/semmle/go/frameworks/`, and a good source of examples. You should make a new file in that folder, named after the library.

## Sources

To mark a source of data that is controlled by an untrusted user, we create a class extending `UntrustedFlowSource::Range`. Inheritance and the characteristic predicate of the class should be used to specify exactly the dataflow node that introduces the data. Here is a short example from `Mux.qll`.

```go
  class RequestVars extends DataFlow::UntrustedFlowSource::Range, DataFlow::CallNode {
    RequestVars() { this.getTarget().hasQualifiedName("github.com/gorilla/mux", "Vars") }
  }
```

## Flow propagation

By default, it is assumed that all functions in libraries do not have any data flow. To indicate that a particular function does, you need to create a class extending `TaintTracking::FunctionModel`. Inheritance and the characteristic predicate of the class should specify the function and a method with the signature `override predicate hasTaintFlow(FunctionInput inp, FunctionOutput outp)` is need. The body should specify either `inp.isReceiver()` or `inp.isParameter(i)` for some `i`. It should also specify either `outp.isResult()` (only valid if there is only one return value), `outp.isResult(i)` or `outp.isParameter(i)` for some `i`. Here is an example from `Gin.qll`, slightly modified for brevity.

```go
private class ParamsGet extends TaintTracking::FunctionModel, Method {
  ParamsGet() { this.hasQualifiedName("github.com/gin-gonic/gin", "Params", "Get") }

  override predicate hasTaintFlow(FunctionInput inp, FunctionOutput outp) {
    inp.isReceiver() and outp.isResult(0)
  }
}
```

## Sanitizers

It is not necessary to indicate that library methods are sanitizers - because the body is not analysed it is assumed that data does not flow through them. 

## Sinks

Data-flow sinks are specified by queries rather than by library models. What library models can do is to indicate when functions belong to special categories of function, which queries can use when specifying sinks. Categories representing these special categories are contained in `ql/src/semmle/go/Concepts.qll`. For example, a call to a logging mechanism should be indicated by making a class that extends `LoggerCall::Range`, as in the following example from `Glog.qll`.

```go
private class GlogCall extends LoggerCall::Range, DataFlow::CallNode {
  GlogCall() {
    exists(string fn |
      fn.regexpMatch("Error(|f|ln)")
      or
      fn.regexpMatch("Exit(|f|ln)")
      or
      fn.regexpMatch("Fatal(|f|ln)")
      or
      fn.regexpMatch("Info(|f|ln)")
      or
      fn.regexpMatch("Warning(|f|ln)")
    |
      this.getTarget().hasQualifiedName("github.com/golang/glog", fn)
      or
      this.getTarget().(Method).hasQualifiedName("github.com/golang/glog", "Verbose", fn)
    )
  }

  override DataFlow::Node getAMessageComponent() { result = this.getAnArgument() }
}
```

Other useful classes in `ql/src/semmle/go/Concepts.qll` include ones for HTTP response writers, HTTP redirects and marshaling and unmarshaling functions.
