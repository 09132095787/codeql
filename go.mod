module github.com/github/codeql-go

go 1.14

require (
	github.com/ChrisTrenkamp/goxpath v0.0.0-20190607011252-c5096ec8773d
	github.com/Masterminds/squirrel v1.4.0
	github.com/lib/pq v1.8.0
	golang.org/x/crypto v0.0.0-20200622213623-75b288015ac9
	golang.org/x/mod v0.3.0
	golang.org/x/oauth2 v0.0.0-20200107190931-bf48bf16ab8d
	golang.org/x/tools v0.0.0-20200827010519-17fd2f27a9e3
)
