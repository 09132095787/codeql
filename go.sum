cloud.google.com/go v0.34.0/go.mod h1:aQUYkXzVsufM+DwF1aE+0xfcU+56JwCaLick0ClmMTw=
github.com/ChrisTrenkamp/goxpath v0.0.0-20190607011252-c5096ec8773d h1:W1diKnDQkXxNDhghdBSbQ4LI/E1aJNTwpqPp3KtlB8w=
github.com/ChrisTrenkamp/goxpath v0.0.0-20190607011252-c5096ec8773d/go.mod h1:nuWgzSkT5PnyOd+272uUmV0dnAnAn42Mk7PiQC5VzN4=
github.com/Masterminds/squirrel v1.4.0 h1:he5i/EXixZxrBUWcxzDYMiju9WZ3ld/l7QBNuo/eN3w=
github.com/Masterminds/squirrel v1.4.0/go.mod h1:yaPeOnPG5ZRwL9oKdTsO/prlkPbXWZlRVMQ/gGlzIuA=
github.com/davecgh/go-spew v1.1.1/go.mod h1:J7Y8YcW2NihsgmVo/mv3lAwl/skON4iLHjSsI+c5H38=
github.com/golang/protobuf v1.2.0 h1:P3YflyNX/ehuJFLhxviNdFxQPkGK5cDcApsge1SqnvM=
github.com/golang/protobuf v1.2.0/go.mod h1:6lQm79b+lXiMfvg/cZm0SGofjICqVBUtrP5yJMmIC1U=
github.com/lann/builder v0.0.0-20180802200727-47ae307949d0 h1:SOEGU9fKiNWd/HOJuq6+3iTQz8KNCLtVX6idSoTLdUw=
github.com/lann/builder v0.0.0-20180802200727-47ae307949d0/go.mod h1:dXGbAdH5GtBTC4WfIxhKZfyBF/HBFgRZSWwZ9g/He9o=
github.com/lann/ps v0.0.0-20150810152359-62de8c46ede0 h1:P6pPBnrTSX3DEVR4fDembhRWSsG5rVo6hYhAB/ADZrk=
github.com/lann/ps v0.0.0-20150810152359-62de8c46ede0/go.mod h1:vmVJ0l/dxyfGW6FmdpVm2joNMFikkuWg0EoCKLGUMNw=
github.com/lib/pq v1.8.0 h1:9xohqzkUwzR4Ga4ivdTcawVS89YSDVxXMa3xJX3cGzg=
github.com/lib/pq v1.8.0/go.mod h1:AlVN5x4E4T544tWzH6hKfbfQvm3HdbOxrmggDNAPY9o=
github.com/pmezard/go-difflib v1.0.0/go.mod h1:iKH77koFhYxTK1pcRnkKkqfTogsbg7gZNVY4sRDYZ/4=
github.com/stretchr/testify v1.2.2/go.mod h1:a8OnRcib4nhh0OaRAV+Yts87kKdq0PP7pXfy6kDkUVs=
github.com/yuin/goldmark v1.1.32/go.mod h1:3hX8gzYuyVAZsxl0MRgGTJEmQBFcNTphYh9decYSb74=
golang.org/x/crypto v0.0.0-20190308221718-c2843e01d9a2/go.mod h1:djNgcEr1/C05ACkg1iLfiJU5Ep61QUkGW8qpdssI0+w=
golang.org/x/crypto v0.0.0-20191011191535-87dc89f01550/go.mod h1:yigFU9vqHzYiE8UmvKecakEJjdnWj3jj499lnFckfCI=
golang.org/x/crypto v0.0.0-20200622213623-75b288015ac9 h1:psW17arqaxU48Z5kZ0CQnkZWQJsqcURM6tKiBApRjXI=
golang.org/x/crypto v0.0.0-20200622213623-75b288015ac9/go.mod h1:LzIPMQfyMNhhGPhUkYOs5KpL4U8rLKemX1yGLhDgUto=
golang.org/x/mod v0.1.1-0.20191105210325-c90efee705ee/go.mod h1:QqPTAvyqsEbceGzBzNggFXnrqF1CaUcvgkdR5Ot7KZg=
golang.org/x/mod v0.2.0 h1:KU7oHjnv3XNWfa5COkzUifxZmxp1TyI7ImMXqFxLwvQ=
golang.org/x/mod v0.2.0/go.mod h1:s0Qsj1ACt9ePp/hMypM3fl4fZqREWJwdYDEqhRiZZUA=
golang.org/x/mod v0.3.0 h1:RM4zey1++hCTbCVQfnWeKs9/IEsaBLA8vTkd0WVtmH4=
golang.org/x/mod v0.3.0/go.mod h1:s0Qsj1ACt9ePp/hMypM3fl4fZqREWJwdYDEqhRiZZUA=
golang.org/x/net v0.0.0-20180724234803-3673e40ba225/go.mod h1:mL1N/T3taQHkDXs73rZJwtUhF3w3ftmwwsq0BUmARs4=
golang.org/x/net v0.0.0-20190108225652-1e06a53dbb7e/go.mod h1:mL1N/T3taQHkDXs73rZJwtUhF3w3ftmwwsq0BUmARs4=
golang.org/x/net v0.0.0-20190404232315-eb5bcb51f2a3/go.mod h1:t9HGtf8HONx5eT2rtn7q6eTqICYqUVnKs3thJo3Qplg=
golang.org/x/net v0.0.0-20190620200207-3b0461eec859/go.mod h1:z5CRVTTTmAJ677TzLLGU+0bjPO0LkuOLi4/5GtJWs/s=
golang.org/x/net v0.0.0-20200625001655-4c5254603344 h1:vGXIOMxbNfDTk/aXCmfdLgkrSV+Z2tcbze+pEc3v5W4=
golang.org/x/net v0.0.0-20200625001655-4c5254603344/go.mod h1:/O7V0waA8r7cgGh81Ro3o1hOxt32SMVPicZroKQ2sZA=
golang.org/x/oauth2 v0.0.0-20200107190931-bf48bf16ab8d h1:TzXSXBo42m9gQenoE3b9BGiEpg5IG2JkU5FkPIawgtw=
golang.org/x/oauth2 v0.0.0-20200107190931-bf48bf16ab8d/go.mod h1:gOpvHmFTYa4IltrdGE7lF6nIHvwfUNPOp7c8zoXwtLw=
golang.org/x/sync v0.0.0-20181221193216-37e7f081c4d4/go.mod h1:RxMgew5VJxzue5/jJTE5uejpjVlOe/izrB70Jof72aM=
golang.org/x/sync v0.0.0-20190423024810-112230192c58/go.mod h1:RxMgew5VJxzue5/jJTE5uejpjVlOe/izrB70Jof72aM=
golang.org/x/sync v0.0.0-20200625203802-6e8e738ad208/go.mod h1:RxMgew5VJxzue5/jJTE5uejpjVlOe/izrB70Jof72aM=
golang.org/x/sys v0.0.0-20190215142949-d0b11bdaac8a/go.mod h1:STP8DvDyc/dI5b8T5hshtkjS+E42TnysNCUPdjciGhY=
golang.org/x/sys v0.0.0-20190412213103-97732733099d/go.mod h1:h1NjWce9XRLGQEsW7wpKNCjG9DtNlClVuFLEZdDNbEs=
golang.org/x/sys v0.0.0-20200323222414-85ca7c5b95cd h1:xhmwyvizuTgC2qz7ZlMluP20uW+C3Rm0FD/WLDX8884=
golang.org/x/sys v0.0.0-20200323222414-85ca7c5b95cd/go.mod h1:h1NjWce9XRLGQEsW7wpKNCjG9DtNlClVuFLEZdDNbEs=
golang.org/x/text v0.3.0 h1:g61tztE5qeGQ89tm6NTjjM9VPIm088od1l6aSorWRWg=
golang.org/x/text v0.3.0/go.mod h1:NqM8EUOU14njkJ3fqMW+pc6Ldnwhi/IjpwHt7yyuwOQ=
golang.org/x/tools v0.0.0-20191119224855-298f0cb1881e/go.mod h1:b+2E5dAYhXwXZwtnZ6UAqBI28+e2cm9otk0dWdXHAEo=
golang.org/x/tools v0.0.0-20200302225559-9b52d559c609 h1:3/QY44rOqJoMLCsQz9bAgInYa08qsu+dH52Uk4DWH3w=
golang.org/x/tools v0.0.0-20200302225559-9b52d559c609/go.mod h1:TB2adYChydJhpapKDTa4BR/hXlZSLoq2Wpct/0txZ28=
golang.org/x/tools v0.0.0-20200827010519-17fd2f27a9e3 h1:r3P/5xOq/dK1991B65Oy6E1fRF/2d/fSYZJ/fXGVfJc=
golang.org/x/tools v0.0.0-20200827010519-17fd2f27a9e3/go.mod h1:njjCfa9FT2d7l9Bc6FUM5FLjQPp3cFF28FI3qnDFljA=
golang.org/x/xerrors v0.0.0-20190717185122-a985d3407aa7/go.mod h1:I/5z698sn9Ka8TeJc9MKroUUfqBBauWjQqLJ2OPfmY0=
golang.org/x/xerrors v0.0.0-20191011141410-1b5146add898 h1:/atklqdjdhuosWIl6AIbOeHJjicWYPqR9bpxqxYG2pA=
golang.org/x/xerrors v0.0.0-20191011141410-1b5146add898/go.mod h1:I/5z698sn9Ka8TeJc9MKroUUfqBBauWjQqLJ2OPfmY0=
golang.org/x/xerrors v0.0.0-20191204190536-9bdfabe68543 h1:E7g+9GITq07hpfrRu66IVDexMakfv52eLZ2CXBWiKr4=
golang.org/x/xerrors v0.0.0-20191204190536-9bdfabe68543/go.mod h1:I/5z698sn9Ka8TeJc9MKroUUfqBBauWjQqLJ2OPfmY0=
google.golang.org/appengine v1.4.0 h1:/wp5JvzpHIxhs/dumFmF7BXTf3Z+dd4uXta4kVyO508=
google.golang.org/appengine v1.4.0/go.mod h1:xpcJRLb0r/rnEns0DIKYYv+WjYCduHsrkT7/EB5XEv4=
