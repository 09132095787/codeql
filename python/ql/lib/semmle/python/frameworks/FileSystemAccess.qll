/**
 * Provides classes modeling security-relevant aspects of the I/O file write or file read operations
 */

private import python
private import semmle.python.dataflow.new.DataFlow
private import semmle.python.dataflow.new.RemoteFlowSources
private import semmle.python.dataflow.new.TaintTracking
private import semmle.python.Concepts
private import semmle.python.ApiGraphs

/**
 * Provides models for the `aiofile` PyPI package.
 * See https://github.com/agronholm/anyio.
 */
private module Aiofile {
  /**
   * A call to the `async_open` function or `AIOFile` constructor from `aiofile` as a sink for Filesystem access.
   */
  class FileResponseCall extends FileSystemAccess::Range, API::CallNode {
    string methodName;

    FileResponseCall() {
      this = API::moduleImport("aiofile").getMember("async_open").getACall() and
      methodName = "async_open"
      or
      this = API::moduleImport("aiofile").getMember("AIOFile").getACall() and
      methodName = "AIOFile"
    }

    override DataFlow::Node getAPathArgument() {
      result = this.getParameter(0, "file_specifier").asSink() and
      methodName = "async_open"
      or
      result = this.getParameter(0, "filename").asSink() and
      methodName = "AIOFile"
    }
  }
}

/**
 * Provides models for the `aiofiles` PyPI package.
 * See https://github.com/Tinche/aiofiles.
 */
private module Aiofiles {
  /**
   * A call to the `open` function from `aiofiles` as a sink for Filesystem access.
   */
  class FileResponseCall extends FileSystemAccess::Range, API::CallNode {
    FileResponseCall() { this = API::moduleImport("aiofiles").getMember("open").getACall() }

    override DataFlow::Node getAPathArgument() { result = this.getParameter(0, "file").asSink() }
  }
}

/**
 * Provides models for the `anyio` PyPI package.
 * See https://github.com/agronholm/anyio.
 */
private module Anyio {
  /**
   * A call to the `from_path` function from `FileReadStream` or `FileWriteStream` constructors of `anyio.streams.file` as a sink for Filesystem access.
   */
  class FileStreamCall extends FileSystemAccess::Range, API::CallNode {
    FileStreamCall() {
      this =
        API::moduleImport("anyio")
            .getMember("streams")
            .getMember("file")
            .getMember(["FileReadStream", "FileWriteStream"])
            .getMember("from_path")
            .getACall()
    }

    override DataFlow::Node getAPathArgument() { result = this.getParameter(0, "path").asSink() }
  }

  /**
   * A call to the `Path` constructor from `anyio` as a sink for Filesystem access.
   */
  class PathCall extends FileSystemAccess::Range, API::CallNode {
    PathCall() { this = API::moduleImport("anyio").getMember("Path").getACall() }

    override DataFlow::Node getAPathArgument() { result = this.getParameter(0).asSink() }
  }

  /**
   * A call to the `open_file` function from `anyio` as a sink for Filesystem access.
   */
  class OpenFileCall extends FileSystemAccess::Range, API::CallNode {
    OpenFileCall() { this = API::moduleImport("anyio").getMember("open_file").getACall() }

    override DataFlow::Node getAPathArgument() { result = this.getParameter(0, "file").asSink() }
  }
}
