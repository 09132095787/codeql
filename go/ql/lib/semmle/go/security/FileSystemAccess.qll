import go

/**
 * The File system access sinks of `net/http` package
 */
class HttpServeFile extends FileSystemAccess::Range, DataFlow::CallNode {
  int pathArg;

  HttpServeFile() {
    exists(Function f |
      f.hasQualifiedName("net/http", "ServeFile") and
      this = f.getACall() and
      pathArg = 2
    )
  }

  override DataFlow::Node getAPathArgument() { result = this.getArgument(pathArg) }
}

/**
 * The File system access sinks of [beego](https://github.com/beego/beego) web framework
 */
class BeegoFileSystemAccess extends FileSystemAccess::Range, DataFlow::CallNode {
  int pathArg;

  BeegoFileSystemAccess() {
    exists(Method m |
      (
        m.hasQualifiedName(package("github.com/beego/beego", "server/web/context"), "BeegoOutput",
          "Download") and
        pathArg = 0
        or
        m.hasQualifiedName(package("github.com/beego/beego", "server/web"), "Controller",
          "SaveToFileWithBuffer") and
        pathArg = 1
      ) and
      this = m.getACall()
    )
  }

  override DataFlow::Node getAPathArgument() { result = this.getArgument(pathArg) }
}

/**
 * Provide File system access sinks of [beego](https://github.com/beego/beego) web framework
 */
class EchoFileSystemAccess extends FileSystemAccess::Range, DataFlow::CallNode {
  int pathArg;

  EchoFileSystemAccess() {
    exists(Method m |
      m.hasQualifiedName(package("github.com/labstack/echo", ""), "Context", ["Attachment", "File"]) and
      this = m.getACall() and
      pathArg = 0
    )
  }

  override DataFlow::Node getAPathArgument() { result = this.getArgument(pathArg) }
}

/**
 * The File system access sinks of [gin](https://github.com/gin-gonic/gin) web framework
 */
class GinFileSystemAccess extends FileSystemAccess::Range, DataFlow::CallNode {
  int pathArg;

  GinFileSystemAccess() {
    exists(Method m |
      (
        m.hasQualifiedName(package("github.com/gin-gonic/gin", ""), "Context",
          ["File", "FileAttachment"]) and
        pathArg = 0
        or
        m.hasQualifiedName(package("github.com/gin-gonic/gin", ""), "Context", "SaveUploadedFile") and
        pathArg = 1
      ) and
      this = m.getACall()
    )
  }

  override DataFlow::Node getAPathArgument() { result = this.getArgument(pathArg) }
}

/**
 * The File system access sinks of [iris](https://github.com/kataras/iris) web framework
 */
class IrisFileSystemAccess extends FileSystemAccess::Range, DataFlow::CallNode {
  int pathArg;

  IrisFileSystemAccess() {
    exists(Method m |
      (
        m.hasQualifiedName(package("github.com/kataras/iris", "context"), "Context",
          ["SendFile", "ServeFile", "SendFileWithRate", "ServeFileWithRate", "UploadFormFiles"]) and
        pathArg = 0
        or
        m.hasQualifiedName(package("github.com/kataras/iris", "context"), "Context", "SaveFormFile") and
        pathArg = 1
      ) and
      this = m.getACall()
    )
  }

  override DataFlow::Node getAPathArgument() { result = this.getArgument(pathArg) }
}

/**
 * The File system access sinks of [fiber](https://github.com/gofiber/fiber) web framework
 */
class FiberSystemAccess extends FileSystemAccess::Range, DataFlow::CallNode {
  int pathArg;

  FiberSystemAccess() {
    exists(Method m |
      (
        m.hasQualifiedName(package("github.com/gofiber/fiber", ""), "Ctx", "SendFile") and
        pathArg = 0
        or
        m.hasQualifiedName(package("github.com/gofiber/fiber", ""), "Ctx", "SaveFile") and
        pathArg = 1
      ) and
      this = m.getACall()
    )
  }

  override DataFlow::Node getAPathArgument() { result = this.getArgument(pathArg) }
}

/**
 * Provide File system access sinks of [afero](https://github.com/spf13/afero) framework
 * The Types that are not vulnerable: `afero.BasePathFs` and `afero.IOFS`
 */
module Afero {
  string aferoPackage() { result = "github.com/spf13/afero" }

  /**
   * Provide File system access sinks of [afero](https://github.com/spf13/afero) framework methods
   */
  class AferoSystemAccess extends FileSystemAccess::Range, DataFlow::CallNode {
    int pathArg;

    AferoSystemAccess() {
      exists(Method f |
        f.hasQualifiedName(package(aferoPackage(), ""), "HttpFs",
          ["Create", "Open", "OpenFile", "Remove", "RemoveAll"]) and
        this = f.getACall() and
        pathArg = 0
        or
        f.hasQualifiedName(package(aferoPackage(), ""), "RegexpFs",
          ["Create", "Open", "OpenFile", "Remove", "RemoveAll", "Mkdir", "MkdirAll"]) and
        this = f.getACall() and
        pathArg = 0
        or
        f.hasQualifiedName(package(aferoPackage(), ""), "ReadOnlyFs",
          ["Create", "Open", "OpenFile", "ReadDir", "ReadlinkIfPossible", "Mkdir", "MkdirAll"]) and
        this = f.getACall() and
        pathArg = 0
        or
        f.hasQualifiedName(package(aferoPackage(), ""), "OsFs",
          [
            "Create", "Open", "OpenFile", "ReadlinkIfPossible", "Remove", "RemoveAll", "Mkdir",
            "MkdirAll"
          ]) and
        this = f.getACall() and
        pathArg = 0
        or
        f.hasQualifiedName(package(aferoPackage(), ""), "MemMapFs",
          ["Create", "Open", "OpenFile", "Remove", "RemoveAll", "Mkdir", "MkdirAll"]) and
        this = f.getACall() and
        pathArg = 0
      )
    }

    override DataFlow::Node getAPathArgument() { result = this.getArgument(pathArg) }
  }

  /**
   * Provide File system access sinks of [afero](https://github.com/spf13/afero) framework utility functions
   * The Types that are not vulnerable: `afero.BasePathFs` and `afero.IOFS`
   */
  class AferoUtilityFunctionSystemAccess extends FileSystemAccess::Range, DataFlow::CallNode {
    int pathArg;

    AferoUtilityFunctionSystemAccess() {
      // utility functions
      exists(Function f |
        f.hasQualifiedName(package(aferoPackage(), ""),
          ["WriteReader", "SafeWriteReader", "WriteFile", "ReadFile", "ReadDir"]) and
        this = f.getACall() and
        pathArg = 1 and
        not aferoSanitizer(this.getArgument(0))
      )
    }

    override DataFlow::Node getAPathArgument() { result = this.getArgument(pathArg) }
  }

  /**
   * A sanitizer for when the Afero utility functions has a first argument of a safe type like NewBasePathFs
   *
   * e.g.
   * ```
   * basePathFs := afero.NewBasePathFs(osFS, "tmp")
   * afero.ReadFile(basePathFs, filepath)
   * ```
   */
  predicate aferoSanitizer(DataFlow::Node n) {
    exists(Function f |
      f.hasQualifiedName(package(aferoPackage(), ""), "NewBasePathFs") and
      TaintTracking::localTaint(f.getACall(), n)
    )
  }

  /**
   * A helper for `aferoSanitizer` for when the Afero instance is initialized with one of the safe FS types like IOFS
   *
   * e.g.`n2 := &afero.Afero{Fs: afero.NewBasePathFs(osFS, "./")}` n1 is `afero.NewBasePathFs(osFS, "./")`
   */
  predicate additionalTaintStep(DataFlow::Node n1, DataFlow::Node n2) {
    exists(StructLit st | st.getType().hasQualifiedName(package(aferoPackage(), ""), "Afero") |
      n1.asExpr() = st.getAChildExpr*() and
      n2.asExpr() = st
    )
  }
}
