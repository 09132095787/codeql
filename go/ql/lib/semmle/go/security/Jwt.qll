/**
 * Provides classes and predicates for reasoning about JSON Web Tokens (JWT).
 */

import go
private import semmle.go.security.HardcodedCredentials

private class IrisJwt extends HardcodedCredentials::Sink {
  IrisJwt() {
    exists(Field f |
      f.hasQualifiedName(package("github.com/kataras/iris", "middleware/jwt"), "Signer", "Key") and
      f.getAWrite().getRhs() = this
    )
  }
}

private class GogfJwtSign extends HardcodedCredentials::Sink {
  GogfJwtSign() {
    exists(Field f |
      f.hasQualifiedName(package("github.com/gogf/gf-jwt", ""), "GfJWTMiddleware", "Key") and
      f.getAWrite().getRhs() = this
    )
  }
}

private class GinJwtSign extends HardcodedCredentials::Sink {
  GinJwtSign() {
    exists(Field f |
      f.hasQualifiedName(package("github.com/appleboy/gin-jwt", ""), "GinJWTMiddleware", "Key") and
      f.getAWrite().getRhs() = this
    )
  }
}

private class SquareJoseKey extends HardcodedCredentials::Sink {
  SquareJoseKey() {
    exists(Field f, string pkg |
      pkg = ["github.com/square/go-jose/v3", "gopkg.in/square/go-jose.v2"]
    |
      f.hasQualifiedName(pkg, ["Recipient", "SigningKey"], "Key") and
      f.getAWrite().getRhs() = this
    )
  }
}
