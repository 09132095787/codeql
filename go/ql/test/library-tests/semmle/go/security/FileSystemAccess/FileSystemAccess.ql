import go
import semmle.go.security.FileSystemAccess

from FileSystemAccess f
select f
