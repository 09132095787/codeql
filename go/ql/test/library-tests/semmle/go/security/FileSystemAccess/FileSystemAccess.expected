| main.go:23:2:23:48 | call to Download |
| main.go:25:6:25:85 | call to SaveToFileWithBuffer |
| main.go:33:14:33:41 | call to MkdirAll |
| main.go:34:14:34:75 | call to WriteFile |
| main.go:35:14:35:77 | call to WriteFile |
| main.go:36:14:36:75 | call to WriteFile |
| main.go:37:16:37:45 | call to ReadFile |
| main.go:39:14:39:32 | call to Open |
| main.go:41:14:41:61 | call to SafeWriteReader |
| main.go:42:14:42:57 | call to WriteReader |
| main.go:48:14:48:47 | call to ReadFile |
| main.go:53:14:53:49 | call to ReadFile |
| main.go:58:14:58:52 | call to ReadFile |
| main.go:63:17:63:37 | call to Open |
| main.go:94:10:94:41 | call to File |
| main.go:98:10:98:72 | call to Attachment |
| main.go:108:7:108:34 | call to SaveFile |
| main.go:109:10:109:29 | call to SendFile |
| main.go:116:6:116:35 | call to SendFile |
| main.go:117:6:117:49 | call to SendFileWithRate |
| main.go:118:6:118:28 | call to ServeFile |
| main.go:119:6:119:42 | call to ServeFileWithRate |
| main.go:120:12:120:45 | call to UploadFormFiles |
| main.go:122:9:122:46 | call to SaveFormFile |
| main.go:129:3:129:18 | call to File |
| main.go:130:3:130:47 | call to ServeFile |
| main.go:131:3:131:53 | call to FileAttachment |
| main.go:133:7:133:40 | call to SaveUploadedFile |
